### PR TITLE
Cortex Xpanse XSOAR Pack

### PR DESCRIPTION
**Pull request by johnnywilkes**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
https://jira-hq.paloaltonetworks.local/browse/EXPANDR-1595

## Description
I will upload a video to explain, please reach out to me for access.
TLDR:
This pack is for XSOAR customers that also have the new version of Cortex Xpanse
https://github.com/demisto/content/tree/master/Packs/ExpanseV2 pack is for XSOAR customer with the legacy version of Xpanse that will be EoL later this year.

## Screenshots
video

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [X] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [X] No

## Must have
- [X] Tests
- [X] Documentation

### CODE REVIEW

**Comment by johnnywilkes** [comment]
I have a video if you need me to share

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 1]
```diff
@@ -0,0 +1,658 @@
...
+            'Authorization': f'{api}',
+            'x-xdr-auth-id': f'{auth_id}',
+            'Content-Type': 'application/json'
+        }
+
+        proxy = params.get('proxy', False)
+        handle_proxy()
+        verify_certificate = not params.get('insecure', False)
+
+        url_suffix = "/public_api/v1"
```
Comment (adi88d) [inline Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:613]: please set `url_suffix` as constant variable and  move it to the top of the code

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+        incidents.append(incident)
+
+        if incident_created_time > latest_created_time:
+            latest_created_time = incident_created_time
+
+    next_run = {'last_fetch': latest_created_time}
+    return next_run, incidents
+
+
+def test_module(client: Client) -> None:
```
Comment (adi88d) [inline Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:558]: in this method check if the integration fetching incidents(`if params.get('isFetch', False):`)
and run also `fetch_incidents` method.

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 3]
```diff
@@ -0,0 +1,658 @@
...
+class Client(BaseClient):
+    """
+    Client class to interact with the service API.
+    """
+
+    def __init__(self, base_url, verify, proxy, headers, auth):
+        """
+        Class initialization.
+        """
+        super().__init__(base_url, verify=verify, proxy=proxy, headers=headers, auth=auth)
```
Comment (adi88d) [inline Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:24]:
```suggestion
        super().__init__(base_url, verify=verify, proxy=proxy, headers=headers)
```

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 4]
```diff
@@ -0,0 +1,658 @@
...
+        url_suffix = "/public_api/v1"
+        url = params.get('url', '')
+        add_sensitive_log_strs(api)
+        base_url = urljoin(url, url_suffix)
+        client = Client(
+            base_url=base_url,
+            verify=verify_certificate,
+            headers=headers,
+            proxy=proxy,
+            auth=None)
```
Comment (adi88d) [inline Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:622]:
```suggestion
            proxy=proxy)
```

File: Packs/CortexXpanse/TestPlaybooks/playbook-CortexXpanse_Test.yml [hunk 5]
```diff
@@ -0,0 +1,961 @@
+id: CortexXpanse_Test
```
Comment (adi88d) [inline Packs/CortexXpanse/TestPlaybooks/playbook-CortexXpanse_Test.yml:1]: in the test playbook no need to Delete context after each command
and please add the script `FetchFromInstance` from `Developer Tools` pack to test the fetch incidents too

**Review by adi88d — CHANGES_REQUESTED** [review]
Hi @johnnywilkes and thank you for your contribution, it looks great!
see my comments below

**Comment by adi88d** [comment]
If you need any help please don’t hesitate to reach out to me here or on Slack

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 6]
```diff
@@ -0,0 +1,658 @@
...
+        formatted_response (list): response from asm-list-asset-internet-exposure command (json)
+
+    Returns:
+        list: list of dictionaries of parsed/formatted json object
+    """
+
+    if formatted_response:
+        for entry in formatted_response:
+            if entry.get('asm_ids'):
+                entry['asm_ids'] = entry['asm_ids'][0]
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:159]: No, that is not possible

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 7]
```diff
@@ -0,0 +1,658 @@
...
+    params: Dict[str, Any] = demisto.params()
+    args: Dict[str, Any] = demisto.args()
+
+    command = demisto.command()
+    demisto.debug(f'Command being called is {command}')
+
+    try:
+        first_fetch_time = arg_to_datetime(
+            arg=params.get('first_fetch', '3 days'),
+            arg_name='First fetch time',
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:594]: complete

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 1 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:613]: complete

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 2 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [adi recommendations](https://github.com/demisto/content/pull/23592/commits/4e5a2f1e787a6dbddfd39f955ab80c0564c8bca0)

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 1 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [adi recommendations](https://github.com/demisto/content/pull/23592/commits/4e5a2f1e787a6dbddfd39f955ab80c0564c8bca0)

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [adi recommendations](https://github.com/demisto/content/pull/23592/commits/4e5a2f1e787a6dbddfd39f955ab80c0564c8bca0)

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 8]
```diff
@@ -0,0 +1,658 @@
...
+    def list_external_service_request(self, search_params: List[Dict]) -> Dict[str, Any]:
+        """Get a list of all your external services using the '/assets/get_external_services/' endpoint.
+
+        Args:
+            search_params (list): list of search parameters to add to the API call body.
+
+        Returns:
+            dict: dict containing list of external services.
+        """
+        data = {"request_data": {"filters": search_params, "search_to": DEFAULT_SEARCH_LIMIT}}
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:52]: @adi88d , I mentioned that this integration is a copy of https://github.com/demisto/content/tree/master/Packs/CortexAttackSurfaceManagement/Integrations/CortexAttackSurfaceManagement and adding fetch and /get_alerts.  Because that behavior is in that integration, i prefer to keep it.

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 9]
```diff
@@ -0,0 +1,658 @@
...
+        severity (str): severity as returned from the ASM API.
+
+    Returns:
+        int: Cortex XSOAR Severity (1 to 4)
+    """
+
+    # In this case the mapping is straightforward, but more complex mappings
+    # might be required in your integration, so a dedicated function is
+    # recommended. This mapping should also be documented.
+    return {
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:181]: @adi88d , i dont really understand the ask here and the why behind it.  Is it possible you would like me to change this to the following?
```
sev_dict = {
    'informational': IncidentSeverity.INFO,
    'low': IncidentSeverity.LOW,
    'medium': IncidentSeverity.MEDIUM,
    'high': IncidentSeverity.HIGH,
    'critical': IncidentSeverity.CRITICAL
}

return sev_dict[severity]
```

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 4 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [Adi recommendation](https://github.com/demisto/content/pull/23592/commits/d733f9f064ec0c341bdc34398bb1a541526edbcb) https://github.com/demisto/content/pull/2

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 3 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [Adi recommendation](https://github.com/demisto/content/pull/23592/commits/d733f9f064ec0c341bdc34398bb1a541526edbcb) https://github.com/demisto/content/pull/2

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 10]
```diff
@@ -0,0 +1,658 @@
...
+        Args:
+            request_data (dict): dict of parameters for API call.
+
+        Returns:
+            dict: dict containing list of external services.
+        """
+        headers = self._headers
+
+        response = self._http_request('POST', '/alerts/get_alerts/',
+                                      json_data=request_data, headers=headers)
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [Adi recommendation](https://github.com/demisto/content/pull/23592/commits/d733f9f064ec0c341bdc34398bb1a541526edbcb) https://github.com/demisto/content/pull/2

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [Adi recommendation](https://github.com/demisto/content/pull/23592/commits/d733f9f064ec0c341bdc34398bb1a541526edbcb) https://github.com/demisto/content/pull/2

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py [hunk 11]
```diff
@@ -0,0 +1,658 @@
...
+                        'search_to': search_to, "sort": {"field": "severity", "keyword": sort_by_severity}}}
+    else:
+        request_data = {"request_data": {"filters": search_params, 'search_from': search_from, 'search_to': search_to}}
+
+    response = client.list_alerts_request(request_data)
+
+    parsed = response.get('reply', {}).get('alerts')
+    markdown = tableToMarkdown('ASM Alerts', parsed, removeNull=True, headerTransform=string_to_table_header)
+    command_results = CommandResults(
+        outputs_prefix='ASM.Alerts',
```
[the remark this answers is not in the document]
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [Adi recommendation](https://github.com/demisto/content/pull/23592/commits/d733f9f064ec0c341bdc34398bb1a541526edbcb) https://github.com/demisto/content/pull/2

File: Packs/CortexXpanse/TestPlaybooks/playbook-CortexXpanse_Test.yml (hunk 5 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/TestPlaybooks/playbook-CortexXpanse_Test.yml:1]: [update test playbook](https://github.com/demisto/content/pull/23592/commits/a1743913e59687fcd43e48f4b5719dafec5597f3)

**Comment by johnnywilkes** [comment]
@adi88d , i followed up on all your suggestion.  let me know what you think, please.

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 2 above)
Comment (adi88d) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: looks good
just change to test the integration always:

client.list_external_service_request([])

if params.get('isFetch'):  # Tests fetch incident:
            severity = params.get('severity')
            max_fetch = int(params.get('max_fetch', 10))
            fetch_incidents(
                client=client,
                max_fetch=max_fetch,
                last_run={},
                first_fetch_time=first_fetch_time,
                severity=severity
            )

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (adi88d) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py:57]: ok

**Comment by adi88d** [comment]
We're ready for a demo. Please check [this page|https://xsoar.pan.dev/docs/contributing/demo-prep], and let me know when you're available for one over DFIR.
thanks

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py (hunk 2 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.py]: [another update to test_module](https://github.com/demisto/content/pull/23592/commits/4c527399388cc61a1b317634421bcac3bd94bba6)

**Comment by efelmandar** [comment]
Hi @johnnywilkes, I went over the content, looks great, looking forward to see it in the demo

**Review by adi88d — APPROVED** [review]
nice!!

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse_description.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+1. Navigate to **Settings** > **Integrations** > **Servers & Services**.
+2. Search for Cortex Xpanse.
+3. Click **Add instance** to create and configure a new integration instance.
+
+    | **Parameter** | **Description** | **Required** |
+    | --- | --- | --- |
+     Server URL | The web UI with \`api-\` appended to front (e.g., https://api-xsiam.paloaltonetworks.com). For more information please see [get-started-with-cortex-xdr-apis](https://docs.paloaltonetworks.com/cortex/cortex-xdr/cortex-xdr-api/cortex-xdr-api-overview/get-started-with-cortex-xdr-apis). | True 
+     API Key ID | See [get-started-with-cortex-xdr-apis](https://docs.paloaltonetworks.com/cortex/cortex-xdr/cortex-xdr-api/cortex-xdr-api-overview/get-started-with-cortex-xdr-apis). | True 
+     API Key | See [get-started-with-cortex-xdr-apis](https://docs.paloaltonetworks.com/cortex/cortex-xdr/cortex-xdr-api/cortex-xdr-api-overview/get-started-with-cortex-xdr-apis). | True
```
Comment (ShirleyDenkberg) [inline Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse_description.md:11]: The yml file does not have the API Key ID or API Key as parameters..
It has 
display: API Key ID
displaypassword: API Key
name: credentials

File: Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.yml [hunk 13]
```diff
@@ -0,0 +1,431 @@
...
+    - description: A date in the format 2019-12-31T23:59:00. Only incidents that were created on or after the specified date/time will be retrieved.
+      name: gte_creation_time
+    - auto: PREDEFINED
+      description: Sorts returned incidents by the date/time that the incident was created ("asc" - ascending, "desc" - descending).
+      name: sort_by_creation_time
+      predefined:
+      - asc
+      - desc
+    - auto: PREDEFINED
+      description: Sorts returned incidents by the date/time that the incident was created ("asc" - ascending, "desc" - descending).
```
Comment (ShirleyDenkberg) [inline Packs/CortexXpanse/Integrations/CortexXpanse/CortexXpanse.yml:381]: I think this should be:
Sorts returned incidents by severity of the incident.

File: Packs/CortexXpanse/Integrations/CortexXpanse/README.md [hunk 14]
```diff
@@ -0,0 +1,1432 @@
...
+#### Input
+
+| **Argument Name** | **Description** | **Required** |
+| --- | --- | --- |
+| alert_id_list | Comma separated list of integers of the alert ID. | Optional | 
+| severity | Comma separated list of strings of alert severity (valid values are low, medium, high, critical, informational). | Optional | 
+| lte_creation_time | A date in the format 2019-12-31T23:59:00. Only incidents that were created on or before the specified date/time will be retrieved. | Optional | 
+| gte_creation_time | A date in the format 2019-12-31T23:59:00. Only incidents that were created on or after the specified date/time will be retrieved. | Optional | 
+| sort_by_creation_time | Sorts returned incidents by the date/time that the incident was created ("asc" - ascending, "desc" - descending). Possible values are: asc, desc. | Optional | 
+| sort_by_severity | Sorts returned incidents by the date/time that the incident was created ("asc" - ascending, "desc" - descending). Possible values are: asc, desc. | Optional |
```
Comment (ShirleyDenkberg) [inline Packs/CortexXpanse/Integrations/CortexXpanse/README.md:1082]: I think this should be:
Sorts returned incidents by severity of the incident.

**Comment by ShirleyDenkberg** [comment]
@adi88d @efelmandar Doc review completed.

File: Packs/CortexXpanse/Integrations/CortexXpanse/README.md (hunk 14 above)
Comment (johnnywilkes) [reply Packs/CortexXpanse/Integrations/CortexXpanse/README.md:1082]: they are alerts in XSIAM/XpanseV2

**Comment by johnnywilkes** [comment]
@adi88d , anything waiting to merge this?
